### PR TITLE
Use more std::unique_ptr and std::make_shared/unique

### PR DESCRIPTION
**Pull request by masterleinad**
The purpose of this PR is twofold

1. In accordance with https://github.com/dealii/dealii/projects/6#card-2223180 try to replace `std::shared_ptr` by `std::unique_ptr` where possible. Especially, in the `TrilinosWrappers` classes, we have a lot of `private` members that can be converted.
2. Replace the `[...]reset(new ...` idiom by `std::make_unique` resp. `std::make_shared`. This is in line with avoiding to call `new` explicitly. This change should also have some minor performance improvements with respect to `std::make_shared`.

To be able to use `std::make_unique` where possible I also had to implement `std_cxx14::make_unique` for unbounded C-style arrays.
This PR does not require any changes in the tests and seems therefore to be mostly backwards compatible. Of course, having `std::unique_ptr` member variables prevents classes from having (default) copy constructors but this PR doesn't remove any.

### CODE REVIEW

File: source/fe/mapping_q.cc [hunk 1]
```diff
@@ -77,7 +77,7 @@ MappingQ<dim,spacedim>::MappingQ (const unsigned int degree,
   // created via the shared_ptr objects
```
Comment (drwells) [inline source/fe/mapping_q.cc]: While you are here: would you mind using `make_shared` on line 74? I only noticed this because of #5883 :)

File: include/deal.II/base/std_cxx14/memory.h [hunk 2]
```diff
@@ -30,15 +30,39 @@ namespace std_cxx14
...
   }
+
+  template <typename T>
+  inline
+  typename std::enable_if<std::is_array<T>::value, std::unique_ptr<T> >::type
+  make_unique(std::size_t n)
+  {
+    static_assert(!internal::is_bounded_array<T>::value,
+                  "This function is not implemented for bounded array types.");
+    return std::unique_ptr<T>(new typename std::remove_extent<T>::type [n]);
```
Comment (drwells) [inline include/deal.II/base/std_cxx14/memory.h]: why won't this work with fixed array lengths?

File: include/deal.II/fe/fe_tools_interpolate.templates.h [hunk 3]
```diff
@@ -111,11 +112,11 @@ namespace FETools
     Vector<typename OutVector::value_type> u2_local(DoFTools::max_dofs_per_cell(dof2));
 
     // have a map for interpolation
-    // matrices. shared_ptr make sure
+    // matrices. unique_ptr make sure
```
Comment (drwells) [inline include/deal.II/fe/fe_tools_interpolate.templates.h]: would you mind fixing the grammar error while you are here?

File: source/base/patterns.cc [hunk 4]
```diff
@@ -755,32 +755,32 @@ namespace Patterns
   {
     if (description.compare(0, std::strlen(description_init), description_init) == 0)
       {
-        int min_elements, max_elements;
+        unsigned int min_elements, max_elements;
```
Comment (drwells) [inline source/base/patterns.cc]: This was already an issue (like most things I found), but should we initialize these values?

File: source/base/patterns.cc [hunk 5]
```diff
@@ -939,7 +939,7 @@ namespace Patterns
...
```
Comment (drwells) [inline source/base/patterns.cc]: same

File: source/lac/trilinos_vector.cc [hunk 6]
```diff
@@ -31,6 +32,7 @@ DEAL_II_DISABLE_EXTRA_DIAGNOSTICS
 DEAL_II_ENABLE_EXTRA_DIAGNOSTICS
 
 #  include <cmath>
+#include <memory>
```
Comment (drwells) [inline source/lac/trilinos_vector.cc]: strange spacing

**Comment by drwells** [comment]
Wow, thats a lot of work! Some of my comments refer to minor (e.g., grammar) problems that were already present. We can leave those for later if you don't want to put them in.

File: source/multigrid/mg_transfer_internal.cc [hunk 7]
```diff
@@ -505,9 +504,7 @@ namespace internal
                 ExcInternalError());
         fe_name[template_starts+1] = '1';
       }
-      std::shared_ptr<FiniteElement<1> > fe_1d
-      (FETools::get_fe_by_name<1,1>(fe_name));
-      const FiniteElement<1> &fe = *fe_1d;
+      const FiniteElement<1> &fe = *(FETools::get_fe_by_name<1,1>(fe_name));
```
[the remark this answers is not in the document]
Comment (bangerth) [reply source/multigrid/mg_transfer_internal.cc]: Yes, this will access invalid data. I think you probably want to just make `fe` a `std::unique_ptr` here.

[the remark this answers is not in the document]
Comment (masterleinad) [reply source/multigrid/mg_transfer_internal.cc]: As far as I know, binding to a local const reference prolongs the lifetime of the object (see https://herbsutter.com/2008/01/01/gotw-88-a-candidate-for-the-most-important-const/).

I am happy to change it to a `std::unique_ptr` anyway.

File: include/deal.II/base/std_cxx14/memory.h (hunk 2 above)
Comment (masterleinad) [reply include/deal.II/base/std_cxx14/memory.h]: As can be seen in https://godbolt.org/g/n6TsgS, `std::unique_ptr<int[7]>` cannot be initialized by a `int *` object since these are different types. The problem is that a `int[7]` variable immediately decays to a `int*`. 
In the end, the standard wants this overload to be deleted anyway. A `static_assert` just seemed easier.

File: source/base/patterns.cc (hunk 4 above)
Comment (masterleinad) [reply source/base/patterns.cc]: I guess initializing doesn't hurt here.

**Comment by masterleinad** [comment]
@drwells I fixed your comments.

**Comment by bangerth** [comment]
Regarding the extension of the lifetime of a temporary: I believe that in a statement like this
```
  const FiniteElement<1> &fe = *(FETools::get_fe_by_name<1,1>(fe_name));
```
the temporary object that results from evaluating *the entire right hand side* has its lifetime extended. This object has type `FiniteElement*`, i.e., it is an automatic variable of pointer type. The problem is that this pointer variable points to data that has been deleted when the temporary `std::unique_ptr` variable was destroyed.

In other words, the temporary whose lifetime is extended is not the `std::unique_ptr`. And so you're in trouble.

**Comment by drwells** [comment]
@bangerth is correct. consider the following program:

```cpp
// life.cc
#include <iostream>
#include <memory>

struct A
{
  A()
  {
    std::cout << "A()\n";
  }

  ~A()
  {
    std::cout << "~A()\n";
  }
};

std::unique_ptr<A> make_a()
{
  return std::make_unique<A>();
}

int main()
{
  std::cout << "started calling make_a\n";
  const A &a = *make_a();
  std::cout << "finished calling make_a\n";
}
```

console output:
```
[drwells@learch const-reference-life]$ g++ -std=c++14 life.cc
[drwells@learch const-reference-life]$ ./a.out 
started calling make_a
A()
~A()
finished calling make_a
```

**Comment by drwells** [comment]
These things are so hard to understand that (like why we mark things as `virtual` in child classes, etc) I think we should avoid them.

**Comment by masterleinad** [comment]
> consider the following program:

Actually, I only see anything reported using `-fsanitize=leak` or `valgrind` for raw pointers but not for `std::unique_ptr`.

Anyway, I switched to assigning to a `std::unique_ptr` which should circumvent this discussion.

File: include/deal.II/base/parameter_handler.h [hunk 8]
```diff
@@ -1627,7 +1627,7 @@ class ParameterHandler : public Subscriptor
  *
  *     void HelperClass::create_new (const unsigned int run_no)
  *     {
- *       p.reset(new Problem());
+ *       p = std::make_shared<Problem());>
```
Comment (bangerth) [inline include/deal.II/base/parameter_handler.h]: I don't think this will compile :-)

File: include/deal.II/base/std_cxx14/memory.h (hunk 2 above)
Comment (bangerth) [reply include/deal.II/base/std_cxx14/memory.h]: The reason why C++ doesn't allow fixed-sized arrays as argument types is that there is no way to initialize them. You can't pass in an array of size 7 as argument to the function -- or better, you can, but you can't copy a whole array. And you also can't pass in 7 separate arguments because they can't be used to copy-construct the elements of an array. In general, initializing arrays of any size that are allocated on the heap is not possible with the exception of initializing them all to the same value.

```
  std::make_shared<int[]>(7);
```
as in the goldbolt example does something different: it allocates an integer array of length 7. But it does not initialize the 7 elements to any given initializer.

File: include/deal.II/base/parameter_handler.h (hunk 8 above)
Comment (masterleinad) [reply include/deal.II/base/parameter_handler.h]: Yes, you're right. :-)

File: include/deal.II/base/std_cxx14/memory.h (hunk 2 above)
Comment (drwells) [reply include/deal.II/base/std_cxx14/memory.h]: This is such a counterintuitive part of the language. I can't figure out how to initialize

```cpp
std::unique_ptr<int[5]> x;
```
and I am not so sure that there is even an answer.

Comment (bangerth) [reply include/deal.II/base/std_cxx14/memory.h]: You can't initialize it. But that's no different from the following:
```
  typedef int five_ints[5];
  five_ints *p = new five_ints;
```
You can allocate such an object, but there is no syntax to initialize the 5 values at the same time as you allocate them, except that
```
  five_ints *p = new five_ints();
```
calls the default constructor for each integer (which sets them to zero).

**Comment by jppelteret** [comment]
@masterleinad This PR now has a conflict that requires some attention.

**Comment by drwells** [comment]
Looks good to me
